frescobaldi_app/extensions/__init__.py

```diff
@@ -0,0 +1,914 @@
+                m.addMenu(entry)
+            else:
+                m.addAction(entry)
+
+        m = self._menus.get(target, None)
+        if not m:
+            # Menu is not cached already, create it
+            actions = self.action_collection().menu_actions(target)
+            if target != 'tools' and not actions:
+                # empty context menu
+                return None
+            panel = self.panel()
+            # An extension without at least one action or panel is invalid
+            if not (panel or actions):
+                m = QMenu(_("Invalid extension menu: {}".format(
+                    self.display_name())))
+                return m
+
+            # finally create the menu
+            m = self._menus[target] = QMenu(
+                self.display_name(), self.mainwindow())
+            if self.has_icon():
+                m.setIcon(self.icon())
+            if panel and target == 'tools':
+                panel.toggleViewAction().setText(_("&Tool Panel"))
+                m.addAction(panel.toggleViewAction())
+                if actions:
+                    m.addSeparator()
+            for entry in actions:
+                add_action_or_submenu(m, entry)
+        return m
+
+    def name(self):
+        """Return the extension's internal name
+        (which actually is retrieved from the directory name)."""
+        return self._name
+
+    def panel(self):
+        """Return the extension's Tool panel, if available, None otherwise."""
+        return self._panel
+
+    def root_directory(self):
+        """Return the extension's root directory."""
+        return self._root_directory
+
+    def set_load_time(self, ms):
+        """Store the time needed to load the application.
+        A string with two fractional digits in 'ms'."""
+        self._load_time = ms
+
+    def settings(self):
+        """Reference to the extension's settings object."""
+        return self._settings
+
+    def settings_changed(self, key, old, new):
+        """Called when a setting *within the extension* has been changed.
+        Complements app_settings_changed.
+        Can be implemented in subclasses."""
+        pass
+
+    def widget(self):
+        """Returns the extension's Tool Panel Widget, if available. or None."""
+        return self.panel().widget() if self.panel() else None
+
+    # Convenience functions to access elements of the application
+    # and the current documents. This will be extended.
+
+    def current_document(self):
+        """Returns the currently opened document or None."""
+        return self.mainwindow().currentDocument()
+
+    def mainwindow(self):
+        """Returns a reference to the main window."""
+        return self.parent().mainwindow()
+
+    def text_cursor(self):
+        """The text cursor of the current document."""
+        return self.mainwindow().textCursor()
+
+
+class ExtensionMixin(QObject):
+    """Convenience mixin class for widgets in an extension.
```

Comment:
@uliska Sorry I don't have the time to dig into the details of the code! Could you reply more about why you are choosing to use both mixins and inheritance?

Reply:
I came to the conclusion that the functionality provided by this class had to be available *anywhere*.  What this approach solves is the problem to have it available in classes that inherit from arbitrary widgets or non-gui classes.

There are some preconfigured classes that can be used in various parts of an extension, but *any* class, be it a widget or any other class can additionally inherit from `ExtensionMixin` to "attach" itself to the extension (usually for direct access to the `settings()` and other properties.

Does that help?

Reply:
Thanks! I have not a clear overview of this (yet). It just seemed to me to be an unusual design choice.